codalab/client/local_bundle_client.py

```diff
@@ -530,21 +511,75 @@ def select_unreadable_worksheets(uuids):
+    def check_download_permission(self, target):
+        check_bundles_have_read_permission(self.model, self._current_user(), [target[0]])
 
     def get_target_info(self, target, depth):
-        check_bundles_have_read_permission(self.model, self._current_user(), [target[0]])
-        path = self.get_target_path(target)
-        if path is None:
-            return None
-        return path_util.get_info(path, depth)
+        """
+        Returns information about an individual target inside the bundle, or
+        None if the target doesn't exist.
+        """
+        self.check_download_permission(target)
+        return self.download_manager.get_target_info(target[0], target[1], depth)
+
+    def open_tarred_gzipped_directory(self, target):
+        """
+        Opens a tarred and gzipped archive of the target directory for
+        streaming. The caller should ensure that the target is a directory.
+        """
+        self.check_download_permission(target)
+        return self.download_manager.stream_tarred_gzipped_directory(target[0], target[1])
+
+    def open_gzipped_file(self, target):
+        """
+        Opens a gzipped archive of the target file. The caller should ensure
+        that the target is a file.
+        """
+        self.check_download_permission(target)
+        return self.download_manager.stream_file(target[0], target[1], gzipped=True)
+
+    def download_directory(self, target, download_path):
+        """
+        Downloads the target directory to the given path. The caller should
+        ensure that the target is a directory.
+        """
+        self.check_download_permission(target)
+        with closing(self.download_manager.stream_tarred_gzipped_directory(target[0], target[1])) as fileobj:
+            os.mkdir(download_path)
+            un_tar_gzip_directory(fileobj, download_path)
+
+    def download_file(self, target, download_path):
+        """
+        Downloads the target file to the given path. The caller should
+        ensure that the target is a file.
+        """
+        self._do_download_file(target, out_path=download_path)
 
     def cat_target(self, target, out):
-        check_bundles_have_read_permission(self.model, self._current_user(), [target[0]])
-        path = self.get_target_path(target)
-        if path is None:
-            return
-        path_util.cat(path, out)
+        """
+        Prints the contents of the target file into the file-like object out.
+        The caller should ensure that the target is a file.
+        """
+        self._do_download_file(target, out_fileobj=out)
+
+    def _do_download_file(self, target, out_path=None, out_fileobj=None):
+        self.check_download_permission(target)
+        with closing(self.download_manager.stream_file(target[0], target[1], gzipped=False)) as fileobj:
+            if out_path is not None:
+                with open(out_path, 'wb') as out:
+                    shutil.copyfileobj(fileobj, out)
+            else:
```

Comment:
Replace this line with 'if out_fileobj is not None' to be semantically more correct